crates/router/src/core/revenue_recovery.rs

```
@@ -284,6 +434,353 @@ pub async fn perform_payments_sync(
     Ok(())
 }
 
+pub async fn perform_calculate_workflow(
+    state: &SessionState,
+    process: &storage::ProcessTracker,
+    tracking_data: &pcr::RevenueRecoveryWorkflowTrackingData,
+    revenue_recovery_payment_data: &pcr::RevenueRecoveryPaymentData,
+    payment_intent: &PaymentIntent,
+) -> Result<(), sch_errors::ProcessTrackerError> {
+    let db = &*state.store;
+    let merchant_id = revenue_recovery_payment_data.merchant_account.get_id();
+    let profile_id = revenue_recovery_payment_data.profile.get_id();
+    let billing_mca_id = revenue_recovery_payment_data.billing_mca.get_id();
+
+    logger::info!(
+        process_id = %process.id,
+        payment_id = %tracking_data.global_payment_id.get_string_repr(),
+        "Starting CALCULATE_WORKFLOW..."
+    );
+
+    // 1. Extract customer_id and token_list from tracking_data
+    let customer_id = payment_intent
+        .extract_connector_customer_id_from_payment_intent()
+        .map_err(|_| sch_errors::ProcessTrackerError::MissingRequiredField)?;
```

don't discard error, 
change context of the error to `MissingRequiredField`